src/python/WMCore/Services/DBS/DBS3Reader.py

```diff
@@ -934,27 +924,51 @@ def fixMissingParentageDatasets(self, childDataset, insertFlag=True):
         :param childDataset: child dataset need to set the parentage correctly.
         :return: blocks which failed to insert parentage. for retry
         """
-        pDatasets = self.listDatasetParents(childDataset)
-        self.logger.info("Parent datasets for %s are: %s", childDataset, pDatasets)
-        # print("parent datasets %s\n" % pDatasets)
-        # pDatasets format is
-        # [{'this_dataset': '/SingleMuon/Run2016D-03Feb2017-v1/MINIAOD', 'parent_dataset_id': 13265209, 'parent_dataset': '/SingleMuon/Run2016D-23Sep2016-v1/AOD'}]
-        if not pDatasets:
+        parentDatasets = self.listDatasetParents(childDataset)
+        self.logger.info("Parent datasets for %s are: %s", childDataset, parentDatasets)
+        # parentDatasets format is
+        # [{'this_dataset': '/SingleMuon/Run2016D-03Feb2017-v1/MINIAOD',
+        #   'parent_dataset_id': 13265209,
+        #   'parent_dataset': '/SingleMuon/Run2016D-23Sep2016-v1/AOD'}]
+        if not parentDatasets:
             self.logger.warning("No parent dataset found for child dataset %s", childDataset)
             return {}
 
-        parentFullInfo = self.getParentDatasetTrio(childDataset)
+        parentFlatData = self.getParentDatasetTrio(childDataset)
+
         blocks = self.listBlocksWithNoParents(childDataset)
         failedBlocks = []
         self.logger.info("Found %d blocks without parentage information", len(blocks))
         for blockName in blocks:
             try:
                 self.logger.info("Fixing parentage for block: %s", blockName)
-                numFiles = self.findAndInsertMissingParentage(blockName, parentFullInfo, insertFlag=insertFlag)
-                self.logger.debug("%s file parentage added for block %s", numFiles, blockName)
+                childFlatData = self.getChildBlockTrio(blockName)
+                listChildParent = []
+
+                # first resolve parentage for all common runLumi pairs between childBlock and parentDataset
+                for runLumi in childFlatData.keys() & parentFlatData.keys():
+                    childFileId = childFlatData[runLumi].fileId
+                    parentFileId = parentFlatData[runLumi].fileId
+                    listChildParent.append([childFileId, parentFileId])
+
+                # then add all run lumi pairs which are missing at the parent Dataset by appending None as parentage information
+                for runLumi in childFlatData.keys() - parentFlatData.keys():
```

Comment:
I like this implementation using sets!
I was wondering if we could make it solid with a unit test though. For that, we would need to separate this block of code in a specific function/method, which would get the childFlatData and parentFlatData as input, and return a list of listChildParent to be extended to the main list (DBSUtils could be a potential place to have it).